File: source/go/internal/types/elements.go
package types

import "encoding/json"

type ChoghadiyaQuality string

const (
	QualityAuspicious   ChoghadiyaQuality = "auspicious"
	QualityInauspicious ChoghadiyaQuality = "inauspicious"
	QualityNeutral      ChoghadiyaQuality = "neutral"
)

var AllChoghadiyaQualities = []ChoghadiyaQuality{
	QualityAuspicious, QualityInauspicious, QualityNeutral,
}

type ElementBase struct {
	Index                int     `json:"index"`
	Name                 string  `json:"name"`
	CompletionPercentage float64 `json:"completionPercentage"`
	EndTime              *JSDate `json:"endTime"`
}

type TithiInfo struct {
	Index                int     `json:"index"`
	Name                 string  `json:"name"`
	Paksha               string  `json:"paksha"`
	Number               int     `json:"number"`
	CompletionPercentage float64 `json:"completionPercentage"`
	EndTime              *JSDate `json:"endTime"`
}

type NakshatraInfo struct {
	Index                int     `json:"index"`
	Name                 string  `json:"name"`
	Pada                 int     `json:"pada"`
	DegreesInNakshatra   float64 `json:"degreesInNakshatra"`
	CompletionPercentage float64 `json:"completionPercentage"`
	EndTime              *JSDate `json:"endTime"`
}

type YogaInfo struct {
	Index                int     `json:"index"`
	Name                 string  `json:"name"`
	CompletionPercentage float64 `json:"completionPercentage"`
	EndTime              *JSDate `json:"endTime"`
}

type KaranaType string

const (
	KaranaFixed   KaranaType = "fixed"
	KaranaMovable KaranaType = "movable"
)

var AllKaranaTypes = []KaranaType{KaranaFixed, KaranaMovable}

type KaranaInfo struct {
	Index                int        `json:"index"`
	Name                 string     `json:"name"`
	Type                 KaranaType `json:"type"`
	CompletionPercentage float64    `json:"completionPercentage"`
	EndTime              *JSDate    `json:"endTime"`
}

type VaraName struct {
	Name  string
	Short string
}

type VaraInfo struct {
	Index       int    `json:"index"`
	Name        string `json:"name"`
	ShortName   string `json:"shortName"`
	EnglishName string `json:"englishName"`
}

type RashiInfo struct {
	Index int    `json:"index"`
	Name  string `json:"name"`
}

type NakshatraIndexInfo struct {
	Index int    `json:"index"`
	Name  string `json:"name"`
}

type ChandraMasaInfo struct {
	Index           int        `json:"index"`
	Name            string     `json:"name"`
	IsAdhika        bool       `json:"isAdhika"`
	System          MasaSystem `json:"system"`
	AmantaIndex     int        `json:"amantaIndex"`
	AmantaName      string     `json:"amantaName"`
	PurnimantaIndex int        `json:"purnimantaIndex"`
	PurnimantaName  string     `json:"purnimantaName"`
}

type SamvatInfo struct {
	VikramSamvat     int    `json:"vikramSamvat"`
	ShakaSamvat      int    `json:"shakaSamvat"`
	VikramSamvatsara string `json:"vikramSamvatsara"`
	ShakaSamvatsara  string `json:"shakaSamvatsara"`
}

type UtcWindow struct {
	StartMs int64 `json:"start"`
	EndMs   int64 `json:"end"`
}

type TimePeriod struct {
	Start      JSDate `json:"start"`
	End        JSDate `json:"end"`
	StartLocal string `json:"startLocal"`
	EndLocal   string `json:"endLocal"`
}

type UnlocalizedChoghadiyaSlot struct {
	StartMs     int64             `json:"start"`
	EndMs       int64             `json:"end"`
	Index       int               `json:"index"`
	Name        string            `json:"name"`
	Quality     ChoghadiyaQuality `json:"quality"`
	QualityName string            `json:"qualityName"`
}

type ChoghadiyaSlot struct {
	Start       JSDate            `json:"start"`
	End         JSDate            `json:"end"`
	Index       int               `json:"index"`
	Name        string            `json:"name"`
	Quality     ChoghadiyaQuality `json:"quality"`
	QualityName string            `json:"qualityName"`
	StartLocal  string            `json:"startLocal"`
	EndLocal    string            `json:"endLocal"`
}

type UnlocalizedChoghadiyaInfo struct {
	Day   []UnlocalizedChoghadiyaSlot `json:"day"`
	Night []UnlocalizedChoghadiyaSlot `json:"night"`
}

type ChoghadiyaInfo struct {
	Day   []ChoghadiyaSlot `json:"day"`
	Night []ChoghadiyaSlot `json:"night"`
}

type UnlocalizedHoraSlot struct {
	StartMs     int64  `json:"start"`
	EndMs       int64  `json:"end"`
	PlanetIndex int    `json:"planetIndex"`
	Planet      string `json:"planet"`
}

type HoraSlot struct {
	Start       JSDate `json:"start"`
	End         JSDate `json:"end"`
	PlanetIndex int    `json:"planetIndex"`
	Planet      string `json:"planet"`
	StartLocal  string `json:"startLocal"`
	EndLocal    string `json:"endLocal"`
}

type UnlocalizedHoraInfo struct {
	Day   []UnlocalizedHoraSlot `json:"day"`
	Night []UnlocalizedHoraSlot `json:"night"`
}

type HoraInfo struct {
	Day   []HoraSlot `json:"day"`
	Night []HoraSlot `json:"night"`
}

type UnlocalizedGowriSlot struct {
	StartMs     int64             `json:"start"`
	EndMs       int64             `json:"end"`
	Index       int               `json:"index"`
	Name        string            `json:"name"`
	Quality     ChoghadiyaQuality `json:"quality"`
	QualityName string            `json:"qualityName"`
}

type GowriSlot struct {
	Start       JSDate            `json:"start"`
	End         JSDate            `json:"end"`
	Index       int               `json:"index"`
	Name        string            `json:"name"`
	Quality     ChoghadiyaQuality `json:"quality"`
	QualityName string            `json:"qualityName"`
	StartLocal  string            `json:"startLocal"`
	EndLocal    string            `json:"endLocal"`
}

type UnlocalizedGowriInfo struct {
	Day   []UnlocalizedGowriSlot `json:"day"`
	Night []UnlocalizedGowriSlot `json:"night"`
}

type GowriInfo struct {
	Day   []GowriSlot `json:"day"`
	Night []GowriSlot `json:"night"`
}

type UnlocalizedDoGhatiSlot struct {
	StartMs     int64             `json:"start"`
	EndMs       int64             `json:"end"`
	Index       int               `json:"index"`
	Name        string            `json:"name"`
	Quality     ChoghadiyaQuality `json:"quality"`
	QualityName string            `json:"qualityName"`
}

type DoGhatiSlot struct {
	Start       JSDate            `json:"start"`
	End         JSDate            `json:"end"`
	Index       int               `json:"index"`
	Name        string            `json:"name"`
	Quality     ChoghadiyaQuality `json:"quality"`
	QualityName string            `json:"qualityName"`
	StartLocal  string            `json:"startLocal"`
	EndLocal    string            `json:"endLocal"`
}

type UnlocalizedDoGhatiInfo struct {
	Day   []UnlocalizedDoGhatiSlot `json:"day"`
	Night []UnlocalizedDoGhatiSlot `json:"night"`
}

type DoGhatiInfo struct {
	Day   []DoGhatiSlot `json:"day"`
	Night []DoGhatiSlot `json:"night"`
}

type UnlocalizedDurMuhurtaPeriod struct {
	StartMs int64           `json:"start"`
	EndMs   int64           `json:"end"`
	Segment DayNightSegment `json:"segment"`
}

type DurMuhurtaPeriod struct {
	Start      JSDate          `json:"start"`
	End        JSDate          `json:"end"`
	StartLocal string          `json:"startLocal"`
	EndLocal   string          `json:"endLocal"`
	Segment    DayNightSegment `json:"segment"`
}

type DayNightSegment string

const (
	SegmentDay   DayNightSegment = "day"
	SegmentNight DayNightSegment = "night"
)

var AllDayNightSegments = []DayNightSegment{SegmentDay, SegmentNight}

type GandaMulaSeverity string

const (
	GandaMulaMild   GandaMulaSeverity = "mild"
	GandaMulaSevere GandaMulaSeverity = "severe"
)

var AllGandaMulaSeverities = []GandaMulaSeverity{GandaMulaMild, GandaMulaSevere}

type GandaMulaInfo struct {
	Active        bool              `json:"active"`
	NakshatraName string            `json:"nakshatraName"`
	Severity      GandaMulaSeverity `json:"severity"`
}

func (g GandaMulaInfo) MarshalJSON() ([]byte, error) {
	if !g.Active {
		return []byte(`{"active":false}`), nil
	}
	type alias GandaMulaInfo
	return json.Marshal(alias(g))
}

type PanchakaType string

const (
	PanchakaRoga    PanchakaType = "roga"
	PanchakaRaja    PanchakaType = "raja"
	PanchakaAgni    PanchakaType = "agni"
	PanchakaChora   PanchakaType = "chora"
	PanchakaMrityu  PanchakaType = "mrityu"
	PanchakaSamanya PanchakaType = "samanya"
)

var AllPanchakaTypes = []PanchakaType{
	PanchakaRoga, PanchakaRaja, PanchakaAgni, PanchakaChora, PanchakaMrityu, PanchakaSamanya,
}

type PanchakaInfo struct {
	Active    bool         `json:"active"`
	Type      PanchakaType `json:"type"`
	Name      string       `json:"name"`
	IsDosha   bool         `json:"isDosha"`
	OnsetVara int          `json:"onsetVara"`
}

func (p PanchakaInfo) MarshalJSON() ([]byte, error) {
	if !p.Active {
		return []byte(`{"active":false}`), nil
	}
	type alias PanchakaInfo
	return json.Marshal(alias(p))
}

type AnandadiYogaInfo struct {
	Index   int               `json:"index"`
	Name    string            `json:"name"`
	Quality ChoghadiyaQuality `json:"quality"`
}

type BhadraLocation string

const (
	BhadraEarth  BhadraLocation = "earth"
	BhadraHeaven BhadraLocation = "heaven"
	BhadraPaatal BhadraLocation = "paatal"
)

var AllBhadraLocations = []BhadraLocation{BhadraEarth, BhadraHeaven, BhadraPaatal}

type UnlocalizedBhadraVasaSegment struct {
	StartMs      int64          `json:"start"`
	EndMs        int64          `json:"end"`
	Location     BhadraLocation `json:"location"`
	LocationName string         `json:"locationName"`
}

type BhadraVasaSegment struct {
	Start        JSDate         `json:"start"`
	End          JSDate         `json:"end"`
	Location     BhadraLocation `json:"location"`
	LocationName string         `json:"locationName"`
	StartLocal   string         `json:"startLocal"`
	EndLocal     string         `json:"endLocal"`
}

type UnlocalizedBhadraInfo struct {
	StartMs      int64                          `json:"start"`
	EndMs        int64                          `json:"end"`
	Location     BhadraLocation                 `json:"location"`
	LocationName string                         `json:"locationName"`
	Vasa         []UnlocalizedBhadraVasaSegment `json:"vasa"`
	IsActive     bool                           `json:"isActive"`
}

type BhadraInfo struct {
	Start        JSDate              `json:"start"`
	End          JSDate              `json:"end"`
	StartLocal   string              `json:"startLocal"`
	EndLocal     string              `json:"endLocal"`
	Location     BhadraLocation      `json:"location"`
	LocationName string              `json:"locationName"`
	Vasa         []BhadraVasaSegment `json:"vasa"`
	IsActive     bool                `json:"isActive"`
}

type SpecialYogaType string

const (
	YogaAmritSiddhi     SpecialYogaType = "amrit_siddhi"
	YogaSarvarthaSiddhi SpecialYogaType = "sarvartha_siddhi"
	YogaRaviPushya      SpecialYogaType = "ravi_pushya"
	YogaGuruPushya      SpecialYogaType = "guru_pushya"
	YogaDwipushkar      SpecialYogaType = "dwipushkar"
	YogaTripushkar      SpecialYogaType = "tripushkar"
	YogaJwalamukhi      SpecialYogaType = "jwalamukhi"
	YogaAadal           SpecialYogaType = "aadal"
	YogaVidaal          SpecialYogaType = "vidaal"
	YogaRavi            SpecialYogaType = "ravi"
)

var AllSpecialYogaTypes = []SpecialYogaType{
	YogaAmritSiddhi, YogaSarvarthaSiddhi, YogaRaviPushya, YogaGuruPushya,
	YogaDwipushkar, YogaTripushkar, YogaJwalamukhi, YogaAadal, YogaVidaal, YogaRavi,
}

type SpecialYogaInfo struct {
	Name string          `json:"name"`
	Type SpecialYogaType `json:"type"`
}

File: source/ts/src/types/elements.ts
/** `startLocal` / `endLocal` are offset-carrying ISO 8601 in the result's timezone. */
export interface TimePeriod {
  start: Date;
  end: Date;
  startLocal: string;
  endLocal: string;
}

export type Unlocalized<T extends TimePeriod> = Omit<T, 'startLocal' | 'endLocal'>;

export type UtcWindow = Unlocalized<TimePeriod>;

/** `segment` names the half the ordinal counts in (15 each). Only Tuesday's second window is a night one. */
export interface DurMuhurtaPeriod extends TimePeriod {
  segment: 'day' | 'night';
}

export interface UnlocalizedInfo<T extends TimePeriod> {
  day: Unlocalized<T>[];
  night: Unlocalized<T>[];
}

interface ElementBase {
  index: number;
  name: string;
  completionPercentage: number;
  endTime: Date | null;
}

export interface TithiInfo extends ElementBase {
  paksha: string;
  number: number;
}

export interface NakshatraInfo extends ElementBase {
  pada: number;
  degreesInNakshatra: number;
}

export type YogaInfo = ElementBase;

export interface KaranaInfo extends ElementBase {
  type: 'fixed' | 'movable';
}

export interface VaraInfo {
  index: number;
  name: string;
  shortName: string;
  englishName: string;
}

/** The first `startTime` is the true boundary instant however far before sunrise; a trailing `endTime` is clamped to next sunrise so the segments tile [sunrise, nextSunrise]. */
interface DailyElementBase {
  startTime: Date | null;
  endTimeLocal: string | null;
  startTimeLocal: string | null;
  isActiveAtSunrise: boolean;
}

export interface DailyTithiInfo extends TithiInfo, DailyElementBase {}
export interface DailyNakshatraInfo extends NakshatraInfo, DailyElementBase {}
export type DailyYogaInfo = YogaInfo & DailyElementBase;
export interface DailyKaranaInfo extends KaranaInfo, DailyElementBase {}

export interface RashiInfo {
  /** 0 = Mesha … 11 = Meena */
  index: number;
  name: string;
}

export interface NakshatraIndexInfo {
  /** 0 = Ashwini … 26 = Revati */
  index: number;
  name: string;
}

export interface ChandraMasaInfo {
  /** 0 = Chaitra … 11 = Phalguna, in the active system */
  index: number;
  name: string;
  /** Leap month: two new moons fall in the same solar month */
  isAdhika: boolean;
  system: 'purnimanta' | 'amanta';
  amantaIndex: number;
  amantaName: string;
  purnimantaIndex: number;
  purnimantaName: string;
}

export type ChoghadiyaQuality = 'auspicious' | 'inauspicious' | 'neutral';

export interface ChoghadiyaSlot extends TimePeriod {
  /** 0-6 within the 7-name Choghadiya cycle */
  index: number;
  name: string;
  quality: ChoghadiyaQuality;
  qualityName: string;
}

export interface ChoghadiyaInfo {
  /** 8 equal slots from sunrise to sunset */
  day: ChoghadiyaSlot[];
  night: ChoghadiyaSlot[];
}

export interface DoGhatiSlot extends TimePeriod {
  /** 0-29 global slot index: 0-14 daytime (Rudra…Bhaga), 15-29 night (Ishwara…Samirana). */
  index: number;
  name: string;
  quality: ChoghadiyaQuality;
  qualityName: string;
}

export interface DoGhatiInfo {
  day: DoGhatiSlot[];
  night: DoGhatiSlot[];
}

export interface GowriSlot extends TimePeriod {
  /** 0-7 within the 8-name Gowri cycle */
  index: number;
  name: string;
  quality: ChoghadiyaQuality;
  qualityName: string;
}

export interface GowriInfo {
  /** 8 equal slots from sunrise to sunset */
  day: GowriSlot[];
  night: GowriSlot[];
}

export interface HoraSlot extends TimePeriod {
  /** 0-6 in Chaldean order: Sun, Venus, Mercury, Moon, Saturn, Jupiter, Mars */
  planetIndex: number;
  planet: string;
}

export interface HoraInfo {
  /** 12 equal horas from sunrise to sunset */
  day: HoraSlot[];
  night: HoraSlot[];
}

export interface SamvatInfo {
  /** Increments at Chaitra Shukla Pratipada */
  vikramSamvat: number;
  /** Same new-year point, 135 years behind Vikram */
  shakaSamvat: number;
  /** 60-year Jovian cycle name */
  vikramSamvatsara: string;
  shakaSamvatsara: string;
}

export interface SpecialYogaInfo {
  name: string;
  type:
    | 'amrit_siddhi'
    | 'sarvartha_siddhi'
    | 'ravi_pushya'
    | 'guru_pushya'
    | 'dwipushkar'
    | 'tripushkar'
    | 'jwalamukhi'
    | 'aadal'
    | 'vidaal'
    | 'ravi';
}

export interface FestivalInfo {
  /** Stable language-independent id; match on this, not the localized `name`. */
  key: string;
  name: string;
  type:
    | 'major'
    | 'minor'
    | 'ekadashi'
    | 'smarta_ekadashi'
    | 'vaishnava_ekadashi'
    | 'pradosha'
    | 'sankranti'
    | 'eclipse';
  description?: string;
}

export type EclipseSubtype = 'partial' | 'total' | 'annular' | 'penumbral';

export interface EclipseInfo {
  kind: 'solar' | 'lunar';
  /** Lunar: by umbral magnitude. Solar: as seen from the location, so a total or annular phase entirely below the horizon reads `'partial'`. */
  subtype: EclipseSubtype;
  start: Date;
  peak: Date;
  end: Date;
  startLocal?: string;
  peakLocal?: string;
  endLocal?: string;
  /** Body above the horizon at peak. */
  visibleFromLocation: boolean;
  /** Fraction of the disc's **area** covered at peak, [0, 1], even when the body is below the horizon then; umbral, so penumbral lunar reads 0. */
  obscuration: number;
  /** Fraction of the **diameter** covered at peak, horizon or not: >1 for a total peak, negative for a penumbral lunar. */
  magnitude: number;
  /** Pre-eclipse impurity window start; null when no sutak applies (penumbral lunar). */
  sutakStart: Date | null;
  /** Umbral last contact (moksha); null for penumbral lunar. */
  sutakEnd: Date | null;
  sutakStartLocal?: string | null;
  sutakEndLocal?: string | null;
  /** Subtype, kind, obscuration and visibility; a solar eclipse whose peak is below the horizon is described at the deepest phase seen, at sunrise or sunset. */
  description: string;
}

/** The vasa follows the Moon's rashi, so a rashi transition splits the Bhadra window. */
export interface BhadraVasaSegment {
  start: Date;
  end: Date;
  startLocal?: string;
  endLocal?: string;
  location: 'earth' | 'heaven' | 'paatal';
  locationName: string;
}

export interface BhadraInfo {
  start: Date;
  end: Date;
  startLocal?: string;
  endLocal?: string;
  /** Keyed on the Moon's rashi **at the window's start**; `vasa` has the breakdown. */
  location: 'earth' | 'heaven' | 'paatal';
  locationName: string;
  /** One segment per Moon rashi the window spans, in time order, tiling [start, end]. */
  vasa: BhadraVasaSegment[];
  isActive: boolean;
}

/**
 * Ashwini (0), Ashlesha (8), Magha (9), Jyeshtha (17), Mula (18), Revati (26), per
 * Muhurta-chintamani / BPHS. Mula and Jyeshtha, spanning the Vrischika/Dhanus boundary, are severe.
 */
export type GandaMulaInfo =
  | { active: false }
  | { active: true; nakshatraName: string; severity: 'mild' | 'severe' };

/** `'samanya'` covers a spell begun on a Wednesday or Thursday, which carries no affliction. */
export type PanchakaType = 'roga' | 'raja' | 'agni' | 'chora' | 'mrityu' | 'samanya';

/** Moon from Dhanishtha 3rd pada through Revati; the type is fixed by the weekday the spell *began* on and holds for the whole spell. */
export type PanchakaInfo =
  | { active: false }
  | {
    active: true;
    type: PanchakaType;
    name: string;
    isDosha: boolean;
    /** 0 = Sunday; the vara that fixed the type. */
    onsetVara: number;
  };

/** The 28-name vara × nakshatra cycle, per Muhurta-chintamani Ch. 4. */
export interface AnandadiYogaInfo {
  /** 0 = Ananda … 27 = Vardhamana. */
  index: number;
  name: string;
  quality: ChoghadiyaQuality;
}
